# Data Wrangling with DataFrames Coding Quiz

Use this Jupyter notebook to find the answers to the quiz in the previous section. There is an answer key in the next part of the lesson.

In [1]:
from pyspark.sql import SparkSession

# TODOS: 
# 1) import any other libraries you might need
# 2) instantiate a Spark session 
# 3) read in the data set located at the path "data/sparkify_log_small.json"
# 4) write code to answer the quiz questions 

from pyspark.sql import SparkSession
from pyspark.sql.functions import udf
from pyspark.sql.types import StringType
from pyspark.sql.types import IntegerType
from pyspark.sql.functions import desc
from pyspark.sql.functions import asc
from pyspark.sql.functions import sum as Fsum

import datetime

import numpy as np
import pandas as pd
%matplotlib inline
import matplotlib.pyplot as plt

In [2]:
spark = SparkSession \
    .builder \
    .appName("Wrangling Data") \
    .getOrCreate()

In [3]:
path = "data/sparkify_log_small.json"
user_log = spark.read.json(path)

# Question 1

Which page did user id "" (empty string) NOT visit?

In [4]:
user_log.groupBy("page").count().show()

+----------------+-----+
|            page|count|
+----------------+-----+
|Submit Downgrade|    1|
|            Home| 1126|
|       Downgrade|   75|
|          Logout|  100|
|   Save Settings|   11|
|           About|   43|
|        Settings|   59|
|           Login|  126|
|        NextSong| 8347|
|            Help|   58|
|         Upgrade|   32|
|           Error|   12|
|  Submit Upgrade|   10|
+----------------+-----+



In [5]:
user_log.where(user_log.userId == "").groupBy("page").count().show()

+-----+-----+
| page|count|
+-----+-----+
| Home|  187|
|About|   15|
|Login|  126|
| Help|    8|
+-----+-----+



TODO: write your code to answer question 1.

Pages not visited by _uderId_ == __""__:
- Submit Downgrade
- Downgrade, Logout
- Save Settings
- Settings
- NextSong
- Upgrade
- Error
- Submit Upgrade

# Question 2 - Reflect

What type of user does the empty string user id most likely refer to?


In [6]:
# TODO: use this space to explore the behavior of the user with an empty string
user_log.where(user_log.userId == "").groupBy("page").count().show()

+-----+-----+
| page|count|
+-----+-----+
| Home|  187|
|About|   15|
|Login|  126|
| Help|    8|
+-----+-----+



A _test user_, employed just to check the behaviour of some basic features on the web page. 

# Question 3

How many female users do we have in the data set?

In [7]:
# TODO: write your code to answer question 3
user_log.dropDuplicates(['userId']).groupBy("gender").count().show()

+------+-----+
|gender|count|
+------+-----+
|     F|  462|
|  null|    1|
|     M|  501|
+------+-----+



There are 462 female users.

# Question 4

How many songs were played from the most played artist?

In [8]:
# TODO: write your code to answer question 4
user_log.groupBy('artist').count().sort('count', ascending=False).show()

+--------------------+-----+
|              artist|count|
+--------------------+-----+
|                null| 1653|
|            Coldplay|   83|
|       Kings Of Leon|   69|
|Florence + The Ma...|   52|
|            Björk|   46|
|       Dwight Yoakam|   45|
|       Justin Bieber|   43|
|      The Black Keys|   40|
|         OneRepublic|   37|
|                Muse|   36|
|        Jack Johnson|   36|
|           Radiohead|   31|
|        Taylor Swift|   29|
|Barry Tuckwell/Ac...|   28|
|          Lily Allen|   28|
|               Train|   28|
|           Daft Punk|   27|
|           Metallica|   27|
|          Nickelback|   27|
|          Kanye West|   26|
+--------------------+-----+
only showing top 20 rows



In [9]:
user_log.select('artist', 'song').where(user_log.artist == 'Coldplay').groupBy('artist').count().show()

+--------+-----+
|  artist|count|
+--------+-----+
|Coldplay|   83|
+--------+-----+



# Question 5 (challenge)

How many songs do users listen to on average between visiting our home page? Please round your answer to the closest integer.



In [10]:
# TODO: write your code to answer question 5